docs/contributing/muzzle.md

```diff
@@ -0,0 +1,32 @@
+# Muzzle
+
+Muzzle is a feature of the Java agent that ensures API compatibility
+between libraries/symbols on the application classpath and APIs of instrumented
+3rd party libraries used by the Agent. In other words the Muzzle ensures
+that the API symbols used by the Agent are compatible with API symbols
+on the application classpath. The Muzzle will prevent loading an instrumentation
+if the APIs do not match.
+
+## How does it work
+
+At build time the Muzzle gradle plugin generates a matcher for 3rd party APIs used by the agent.
+The matcher is stored in the instrumentation class in method `ReferenceMatcher getInstrumentationMuzzle()`.
```

Comment:
Can you add a sentence explaining where do the symbols come from?

```suggestion
At build time, for each instrumentation the Muzzle ByteBuddy plugin collects symbols referring to both internal and 3rd party APIs used by the currently processed instrumentation. The reference collection process starts from advice classes - values of the map returned by the `Instrumenter.Default#transformers()` method.
All those references are then used to create a `ReferenceMatcher` instance.
The matcher is stored in the instrumentation class in method `ReferenceMatcher getInstrumentationMuzzle()`.
```

Reply:
this is nice, i'm not sure about "internal" APIs though, we do currently, but hopefully we won't after #1233?

Reply:
Hopefully we'll get rid of at least most of these